File: backend/oauth_google.py
"""Login social com Google (OAuth 2.0 Authorization Code flow), usando só a
biblioteca padrão do Python para as chamadas HTTP — mesma filosofia de
security.py e totp.py, evitando dependências como authlib/google-auth-library
que têm superfícies grandes e trazem outras dependências transitivas junto.
"""
import json
import os
import urllib.error
import urllib.parse
import urllib.request

GOOGLE_CLIENT_ID = os.getenv("GOOGLE_CLIENT_ID")
GOOGLE_CLIENT_SECRET = os.getenv("GOOGLE_CLIENT_SECRET")
GOOGLE_REDIRECT_URI = os.getenv("GOOGLE_REDIRECT_URI", "http://localhost:8000/api/auth/google/callback")

AUTORIZACAO_URL = "https://accounts.google.com/o/oauth2/v2/auth"
TOKEN_URL = "https://oauth2.googleapis.com/token"
USERINFO_URL = "https://www.googleapis.com/oauth2/v3/userinfo"


def configurado() -> bool:
    return bool(GOOGLE_CLIENT_ID and GOOGLE_CLIENT_SECRET)


def url_autorizacao(state: str) -> str:
    params = {
        "client_id": GOOGLE_CLIENT_ID,
        "redirect_uri": GOOGLE_REDIRECT_URI,
        "response_type": "code",
        "scope": "openid email profile",
        "state": state,
        "prompt": "select_account",
    }
    return f"{AUTORIZACAO_URL}?{urllib.parse.urlencode(params)}"


def _post_form(url: str, dados: dict) -> dict:
    corpo = urllib.parse.urlencode(dados).encode()
    requisicao = urllib.request.Request(url, data=corpo, method="POST")
    requisicao.add_header("Content-Type", "application/x-www-form-urlencoded")
    with urllib.request.urlopen(requisicao, timeout=10) as resposta:
        return json.loads(resposta.read().decode())


class ErroTrocaGoogle(Exception):
    pass


def trocar_codigo_por_perfil(code: str) -> dict:
    """Troca o código de autorização por um access_token e busca o perfil do
    usuário (email, nome, id do Google). Levanta ErroTrocaGoogle se o código
    for inválido/expirado ou a chamada à API do Google falhar."""
    try:
        tokens = _post_form(TOKEN_URL, {
            "client_id": GOOGLE_CLIENT_ID,
            "client_secret": GOOGLE_CLIENT_SECRET,
            "code": code,
            "redirect_uri": GOOGLE_REDIRECT_URI,
            "grant_type": "authorization_code",
        })
        access_token = tokens["access_token"]

        requisicao = urllib.request.Request(USERINFO_URL)
        requisicao.add_header("Authorization", f"Bearer {access_token}")
        with urllib.request.urlopen(requisicao, timeout=10) as resposta:
            return json.loads(resposta.read().decode())
    except (urllib.error.URLError, KeyError, json.JSONDecodeError) as erro:
        raise ErroTrocaGoogle(str(erro)) from erro

File: backend-php/lib/oauth_google.php
<?php
declare(strict_types=1);

// Porta de backend/oauth_google.py (OAuth 2.0 Authorization Code flow), usando cURL
// (extensão padrão do PHP em qualquer hospedagem, inclusive Hostinger).

const GOOGLE_AUTORIZACAO_URL = 'https://accounts.google.com/o/oauth2/v2/auth';
const GOOGLE_TOKEN_URL = 'https://oauth2.googleapis.com/token';
const GOOGLE_USERINFO_URL = 'https://www.googleapis.com/oauth2/v3/userinfo';

class ErroTrocaGoogle extends Exception {}

function google_configurado(): bool
{
    return (bool) (config('GOOGLE_CLIENT_ID') && config('GOOGLE_CLIENT_SECRET'));
}

function google_redirect_uri(): string
{
    return config('GOOGLE_REDIRECT_URI', 'http://localhost:8000/api/auth/google/callback');
}

function google_url_autorizacao(string $state): string
{
    $params = [
        'client_id' => config('GOOGLE_CLIENT_ID'),
        'redirect_uri' => google_redirect_uri(),
        'response_type' => 'code',
        'scope' => 'openid email profile',
        'state' => $state,
        'prompt' => 'select_account',
    ];
    return GOOGLE_AUTORIZACAO_URL . '?' . http_build_query($params);
}

function curl_json(string $url, ?array $postFields = null, ?string $bearer = null): array
{
    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 10);
    if ($postFields !== null) {
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($postFields));
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/x-www-form-urlencoded']);
    }
    if ($bearer !== null) {
        curl_setopt($ch, CURLOPT_HTTPHEADER, ["Authorization: Bearer {$bearer}"]);
    }
    $resposta = curl_exec($ch);
    $erro = curl_error($ch);
    curl_close($ch);
    if ($resposta === false) {
        throw new ErroTrocaGoogle($erro);
    }
    $dados = json_decode($resposta, true);
    if (!is_array($dados)) {
        throw new ErroTrocaGoogle('Resposta inválida do Google');
    }
    return $dados;
}

function google_trocar_codigo_por_perfil(string $code): array
{
    $tokens = curl_json(GOOGLE_TOKEN_URL, [
        'client_id' => config('GOOGLE_CLIENT_ID'),
        'client_secret' => config('GOOGLE_CLIENT_SECRET'),
        'code' => $code,
        'redirect_uri' => google_redirect_uri(),
        'grant_type' => 'authorization_code',
    ]);
    if (!isset($tokens['access_token'])) {
        throw new ErroTrocaGoogle('Google não retornou access_token');
    }
    return curl_json(GOOGLE_USERINFO_URL, null, $tokens['access_token']);
}
